components/push/ffi/src/lib.rs

```diff
@@ -0,0 +1,224 @@
+    let db_path = ffi_support::opt_rust_string_from_c(database_path);
+    let config = PushConfiguration {
+        server_host: host,
+        http_protocol: protocol,
+        bridge_type: bridge,
+        registration_id: reg_id,
+        sender_id: sender,
+        database_path: db_path,
+        ..Default::default()
+    };
+    MANAGER.insert_with_result(error, || PushManager::new(config.clone()))
+}
+
+// Add a subscription
+/// Errors are logged.
+#[no_mangle]
+pub unsafe extern "C" fn push_subscribe(
+    handle: u64,
+    channel_id: *const c_char,
+    scope: *const c_char,
+    error: &mut ExternError,
+) -> *mut c_char {
+    log::debug!("push_get_subscription");
+    MANAGER.call_with_result_mut(error, handle, |mgr| -> Result<String> {
+        let channel = ffi_support::rust_str_from_c(channel_id);
+        let scope_s = ffi_support::rust_str_from_c(scope);
+        // Don't auto add the subscription to the db.
+        // (endpoint updates also call subscribe and should be lighter weight)
+        let (info, subscription_key) = mgr.subscribe(channel, scope_s)?;
+        // store the channelid => auth + subscription_key
+        let subscription_info = json!({
+            "endpoint": info.endpoint,
+            "keys": {
+                "auth": base64::encode_config(&subscription_key.auth,
+                                              base64::URL_SAFE_NO_PAD),
+                "p256dh": base64::encode_config(&subscription_key.public,
+                                                base64::URL_SAFE_NO_PAD)
+            }
+        });
+        return Ok(subscription_info.to_string());
+    })
+}
+
+// Unsubscribe a channel
+#[no_mangle]
+pub unsafe extern "C" fn push_unsubscribe(
+    handle: u64,
+    channel_id: *const c_char,
+    error: &mut ExternError,
+) -> u8 {
+    log::debug!("push_unsubscribe");
+    MANAGER.call_with_result_mut(error, handle, |mgr| -> Result<bool> {
+        let channel = ffi_support::opt_rust_str_from_c(channel_id);
+        mgr.unsubscribe(channel)
+    })
+}
+
+// Update the OS token
+#[no_mangle]
+pub unsafe extern "C" fn push_update(
+    handle: u64,
+    new_token: *const c_char,
+    error: &mut ExternError,
+) -> u8 {
+    log::debug!("push_update");
+    MANAGER.call_with_result_mut(error, handle, |mgr| -> Result<_> {
+        if let Some(token) = ffi_support::opt_rust_str_from_c(new_token) {
+            mgr.update(&token)
+        } else {
+            Err(push_errors::ErrorKind::MissingRegistrationTokenError.into())
+        }
+    })
+}
+
+// verify connection using channel list
+// Returns a JSON containing the new channelids => endpoints
+// NOTE: AC should notify processes associated with channelIDs of new endpoint
+#[no_mangle]
+pub unsafe extern "C" fn push_verify_connection(
+    handle: u64,
+    error: &mut ExternError,
+) -> *mut c_char {
+    log::debug!("push_verify");
+    MANAGER.call_with_result_mut(error, handle, |mgr| -> Result<_> {
```

Comment:
If this returns `Result<Option<WhateverRegenerateEndpointsReturns>>` ffi_support will do a lot of this for you (it will come over as a null pointer on None)

Reply:
I still think this should be done in the same way the other FFIs do it but it's not a huge deal.